### PR TITLE
Eng 13113 fix non restartable sysprocs

### PR DESCRIPTION
**Pull request by luyangco**
Pro: https://github.com/VoltDB/pro/pull/2790

### CODE REVIEW

**Comment by zeemanhe** [comment]
A thought: could the fault parameters be passed in as system or env properties and cached in SysprocFaultInjection.  Thus no VoltDB interface changes are needed for these testing parameters.

File: src/frontend/org/voltdb/sysprocs/SwapTablesCore.java [hunk 1]
```diff
@@ -54,7 +57,18 @@ public DependencyPair executePlanFragment(Map<Integer, List<VoltTable>> dependen
         if (fragmentId == SysProcFragmentId.PF_swapTables) {
             // issue the callback once on each node
             if (context.isLowestSiteId()) {
+                UndoAction undoSwapDR = new UndoAction() {
+                    @Override
+                    public void release() {}
+
+                    @Override
+                    public void undo() {
+                        VoltDB.instance().swapTables((String) params.getParam(0), (String) params.getParam(1));
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/sysprocs/SwapTablesCore.java]: Is there a test which confirms that  this UndoAction will undo the changes in EE?

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 2]
```diff
@@ -403,6 +403,9 @@
     // Using Boolean so if this is accessed before assignment the caller will get an NPE
     private Boolean m_eligibleAsLeader = null;
 
+    // Test only field, simulate different types of failures in system procedures.
+    private List<SysprocFaultInjection> m_faults;
```
Comment (n3world) [inline src/frontend/org/voltdb/RealVoltDB.java]: Is there a way to do this without embedding testing code in the product?

File: src/frontend/org/voltdb/SysprocFaultInjection.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.voltdb;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class SysprocFaultInjection {
```
Comment (n3world) [inline src/frontend/org/voltdb/SysprocFaultInjection.java]: Shouldn't this be a class in the test sources? We don't want fault injection in the running product

File: src/frontend/org/voltdb/sysprocs/UpdateSettings.java [hunk 4]
```diff
@@ -78,21 +80,20 @@ public DependencyPair executePlanFragment(
...
 
         if (fragmentId == SysProcFragmentId.PF_updateSettingsBarrier) {
-
             DependencyPair success = new DependencyPair.TableDependencyPair(SysProcFragmentId.PF_updateSettingsBarrier,
                     new VoltTable(new ColumnInfo[] { new ColumnInfo("UNUSED", VoltType.BIGINT) } ));
             if (log.isInfoEnabled()) {
                 log.info("Site " + CoreUtils.hsIdToString(m_site.getCorrespondingSiteId()) +
                         " reached settings update barrier.");
             }
+            SysprocFaultInjection.check(FaultType.UpdateSettingsFirstFragment);
```
Comment (n3world) [inline src/frontend/org/voltdb/sysprocs/UpdateSettings.java]: This seems like a very intrusive testing hook

File: src/frontend/org/voltdb/RealVoltDB.java (hunk 2 above)
Comment (luyangco) [reply src/frontend/org/voltdb/RealVoltDB.java]: Thanks for bringing the challenge to me! I did some research on this topic and found a neat technique, Byteman is a JVM level bytecode manipulation tool/library that is able to inject fault in the java runtime without adding intrusive code to product, and most importantly it can integrate with JUnit. But the documentation seems a little bit overwhelming plentiful, I need time to absorb. I'll update the branch with the new testing method soon.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java [hunk 5]
```diff
@@ -386,7 +394,7 @@ Procedure asCatalogProcedure() {
                 new Config("org.voltdb.sysprocs.BalancePartitions",
                         false, false, false, 0, VoltType.INVALID,
                         true, false, false, true,
-                        false, true, false));
+                        false, true, true));
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/SystemProcedureCatalog.java]: Since you are here anyway, can you change the last param type for restartable and durable to enum?

File: src/frontend/org/voltdb/iv2/MpProcedureTask.java [hunk 6]
```diff
@@ -124,14 +124,8 @@ public void run(SiteProcedureConnection siteConnection)
         final SystemProcedureCatalog.Config sysproc = SystemProcedureCatalog.listing.get(spName);
 
         // certain system procs can and can't be restarted
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/iv2/MpProcedureTask.java]: For the modified comment below can or can't they be restarted. This line makes that ambiguous.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java [hunk 7]
```diff
@@ -199,10 +200,17 @@ Config build() {
...
         public final boolean restartable;
+        // MP Durable and restartable yes
+        // MP not durable and restartable yes
+        // MP durable and not restartable no
+        // MP not durable and not restartable yes
 
         Config(String className, boolean singlePartition, boolean readOnly, boolean everySite, int partitionParam,
                 VoltType partitionParamType, boolean commercial, boolean terminatesReplication, boolean allowedInReplica,
                 boolean durable, boolean allowedInShutdown, boolean transactional, boolean restartable) {
+            Preconditions.checkArgument(!transactional || singlePartition || restartable || !durable,
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/SystemProcedureCatalog.java]: I don't think this is right. A single partition procedure that is durable but not restartable will not corrupt the command log but will corrupt the recovered state. If the order of execution is different between the original path and the replay path, the final state after recovery will be different as well. A LoadSinglepartitionTable that fails because it was not restarted will be resubmitted by the client and will succeed but in replay the load will succeed and the resubmitted will fail. This will potentially result in a different end state.

File: src/frontend/org/voltdb/iv2/MpProcedureTask.java [hunk 8]
```diff
@@ -124,14 +124,8 @@ public void run(SiteProcedureConnection siteConnection)
...
-        // Right now this is adhoc, catalog update, load MP table, and apply binary log MP.
-        //
-        // Note that we don't restart @BalancePartitions transactions, because they do
-        // partition to master HSID lookups in the run() method. When transactions are
-        // restarted, the run() method is not rerun. Let the elastic join coordinator reissue it.
-        if (m_isRestart &&
-                sysproc != null &&
-                !sysproc.isRestartable())
+        // Right now these are swap table, update logging and load MP table.
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/iv2/MpProcedureTask.java]: I don't think this list is correct any more.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java (hunk 7 above)
Comment (JadenW) [reply src/frontend/org/voltdb/SystemProcedureCatalog.java]: Single partition procedure would not be restart, only rollback on unsuccessful. 
So this restartable flag seems meaningless w.r.t to Single Partition Procedures.

Comment (luyangco) [reply src/frontend/org/voltdb/SystemProcedureCatalog.java]: Restartable flag is useless for SP system procedures because it's used in MpProcedureTask and Mp repair only. If LoadSinglepartitionTable failed due to partition leader crash then new SP leader may or may not re-execute the procedure, depends on whether or not the initiate message arrives at any replica. But all of above are expected and client would get an ambiguous response message claims that the transaction may or may not executed.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java (hunk 5 above)
Comment (luyangco) [reply src/frontend/org/voltdb/SystemProcedureCatalog.java]: Do you mean combine restartable and durable into an enum? What can we benefit from doing that?

File: src/frontend/org/voltdb/iv2/MpProcedureTask.java (hunk 8 above)
Comment (luyangco) [reply src/frontend/org/voltdb/iv2/MpProcedureTask.java]: Removed.

File: src/frontend/org/voltdb/iv2/MpProcedureTask.java (hunk 6 above)
Comment (luyangco) [reply src/frontend/org/voltdb/iv2/MpProcedureTask.java]: Comment updated.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java [hunk 9]
```diff
@@ -281,378 +285,378 @@ Procedure asCatalogProcedure() {
...
                         false, false, false, 0, VoltType.INVALID,
                         false, false, true, false,
-                        true, true, false));
+                        true, true, Restartability.NON_RESTARTABLE));
         builder.put("@SwapTables",
                 new Config("org.voltdb.sysprocs.SwapTables",
                         false, false, false, 0,    VoltType.INVALID,
                         false, false, true, true,
-                        false, false, false));
+                        false, false, Restartability.NOT_APPLICABLE));
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/SystemProcedureCatalog.java]: Isn't this a durable MP Procedure?

File: src/frontend/org/voltdb/SystemProcedureCatalog.java [hunk 10]
```diff
@@ -281,378 +285,378 @@ Procedure asCatalogProcedure() {
...
                         false, false, false, 0, VoltType.VARBINARY,
                         false, false, false, true,
-                        false, true, true));
+                        false, true, Restartability.RESTARTABLE));
         builder.put("@MigrateRowsNT",
                 new Config("org.voltdb.sysprocs.MigrateRowsNT",
                         true, false, false, 0, VoltType.INVALID,
                         false, false, false, true,
-                        false, false, false));
+                        false, false, Restartability.NOT_APPLICABLE));
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/SystemProcedureCatalog.java]: Since this is an NT proc, it is not durable either right? Maybe we need an enum for durable as well to make this clear.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java (hunk 9 above)
Comment (luyangco) [reply src/frontend/org/voltdb/SystemProcedureCatalog.java]: Right, I updated the boolean to an enum.

File: src/frontend/org/voltdb/SystemProcedureCatalog.java (hunk 10 above)
Comment (luyangco) [reply src/frontend/org/voltdb/SystemProcedureCatalog.java]: Right, I updated the boolean to an enum.
